### PR TITLE
SPARKC-686 scala 2.13 support

### PR DESCRIPTION
**Pull request by SamTheisens**
# Description
## How did the Spark Cassandra Connector Work or Not Work Before this Patch
Spark Cassandra Connector did not work with Scala 2.13 before this patch.

## General Design of the patch
This is another attempt inspired by https://github.com/datastax/spark-cassandra-connector/pull/1349. The approach has been as follows:

 - Applied automatated scala collection migration via [scalafix](https://github.com/scala/scala-collection-compat)
 - Conditionally add dependencies on `org.scala-lang.modules.scala-collection-compat` and `org.scala-lang.modules.scala-parallel-collections` for scala 2.13
 - Scala version specific implementations of `SparkILoop`, `ParIterable`, `GenericJavaRowReaderFactory` and `CanBuildFrom`.

# How Has This Been Tested?
All automated tests have [passed](https://github.com/SamTheisens/spark-cassandra-connector/actions/runs/5529153873) in my [fork](https://github.com/SamTheisens/spark-cassandra-connector/pull/1), but the `(2.12.11, 3.11.10)` build seems to time out intermittently at cluster startup:
```
java.lang.RuntimeException: The command [ccm, create, ccm_1, -i, 127.12.0., -v, 5.1.24, --dse, --config-dir=/tmp/ccm6059075771339504008] was killed after 10 minutes
```

# Checklist:
* [x]  I have a ticket in the [OSS JIRA](https://datastax-oss.atlassian.net/browse/SPARKC-686)
* [x]  I have performed a self-review of my own code
* [x]  Locally all tests pass (make sure tests fail without your patch)

### CODE REVIEW

**Comment by SamTheisens** [comment]
@jtgrabowski  The intermittence of the [failing tests](https://github.com/datastax/spark-cassandra-connector/pull/1361/files#annotation_12596311180) doesn't seem to depend on the particular version of scala or cassandra used. Rather, I seem to have more success in the early hours in my timezone (Indonesia).
Re-running failed tests only tends to get all the lights green.

Could it be that we run into timing problems due to varying amounts of resources available to github actions at different times of the day? 
Screenshot 2023-07-14 at 15 44 27

**Comment by jtgrabowski** [comment]
I don't know what is causing this, the errors seam unrelated to the PR.
I apologize, but I won't be able to take a look until 24th of July. Let's revisit then.

**Comment by SamTheisens** [comment]
> I don't know what is causing this, the errors seam unrelated to the PR. I apologize, but I won't be able to take a look until 24th of July. Let's revisit then.

No worries! Anything I could do in the mean time to help preparing a new release?

File: connector/src/it/scala/com/datastax/spark/connector/rdd/RDDSpec.scala [hunk 1]
```diff
@@ -12,7 +11,9 @@ import com.datastax.spark.connector.cql.CassandraConnector
 import com.datastax.spark.connector.embedded.SparkTemplate._
 import com.datastax.spark.connector.rdd.partitioner.EndpointPartition
 import com.datastax.spark.connector.writer.AsyncExecutor
+import spire.ClassTag
```
Comment (jtgrabowski) [inline connector/src/it/scala/com/datastax/spark/connector/rdd/RDDSpec.scala:14]: Use scala.reflect.ClassTag

File: connector/src/it/scala/com/datastax/spark/connector/writer/TableWriterSpec.scala [hunk 2]
```diff
@@ -771,7 +771,7 @@ class TableWriterSpec extends SparkCassandraITFlatSpecBase with DefaultCluster {
     verifyKeyValueTable("key_value")
   }
 
-  it should "be able to append and prepend elements to a C* list" in {
+  it should "be able to.append and.prepend elements to a C* list" in {
```
Comment (jtgrabowski) [inline connector/src/it/scala/com/datastax/spark/connector/writer/TableWriterSpec.scala:774]: Could you remove dots from test name here and others below?

File: driver/src/main/scala/com/datastax/spark/connector/mapper/GettableDataToMappedTypeConverter.scala [hunk 3]
```diff
@@ -175,7 +175,7 @@ class GettableDataToMappedTypeConverter[T : TypeTag : ColumnMapper](
       for ((s, _) <- columnMap.setters)
         yield (s, ReflectionUtil.methodParamTypes(targetType, s).head)
     val setterColumnTypes: Map[String, ColumnType[_]] =
-      columnMap.setters.mapValues(columnType)
+      columnMap.setters.map{case (k, v) => (k, columnType(v))}.toMap
```
Comment (jtgrabowski) [inline driver/src/main/scala/com/datastax/spark/connector/mapper/GettableDataToMappedTypeConverter.scala:178]: Just curious, why did you choose `.map().toMap` over `.view.mapValues...`?

**Review by jtgrabowski — COMMENTED** [review]
This is great. Left minor comments.

File: connector/src/it/scala/com/datastax/spark/connector/rdd/RDDSpec.scala (hunk 1 above)
Comment (SamTheisens) [reply connector/src/it/scala/com/datastax/spark/connector/rdd/RDDSpec.scala]: Fixed in [[SPARKC-686] Fix incorrect import](https://github.com/datastax/spark-cassandra-connector/pull/1361/commits/cb36121b0f6c6142153b94dc09f19659ac26ffbe)

File: connector/src/it/scala/com/datastax/spark/connector/writer/TableWriterSpec.scala (hunk 2 above)
Comment (SamTheisens) [reply connector/src/it/scala/com/datastax/spark/connector/writer/TableWriterSpec.scala]: Fixed in [[SPARKC-686] Remove accidental dots in spec names](https://github.com/datastax/spark-cassandra-connector/pull/1361/commits/c6d84c6b395f343903dc7d61aec3eb99da966885)

File: driver/src/main/scala/com/datastax/spark/connector/mapper/GettableDataToMappedTypeConverter.scala (hunk 3 above)
Comment (SamTheisens) [reply driver/src/main/scala/com/datastax/spark/connector/mapper/GettableDataToMappedTypeConverter.scala]: IIRC I was struggling to find a syntax that works in both `2.12 and `2.13`. It looks like `mapValues` was actually fine as long as the return type is converted back to a map.
[[SPARKC-686] Replace closure with idiomatic mapValues](https://github.com/datastax/spark-cassandra-connector/pull/1361/commits/1e5702e3acd4d5f5f01c5c3d8cd8e090a7c052a7)

**Review by jtgrabowski — APPROVED** [review]
LGTM! Thank you for this awesome contribution.
Please sing CLA (https://cla.datastax.com/) if you haven't already.

**Comment by jtgrabowski** [comment]
Thank you @SamTheisens ! A new scc version should be released next week.

**Comment by SamTheisens** [comment]
> Thank you @SamTheisens ! A new scc version should be released next week.

@jtgrabowski anything I could help out with towards creating a release?

**Comment by SamTheisens** [comment]
> @SamTheisens the release is now done. The artifacts should appear in the public repository shortly. Thanks again!

Awesome! Thanks a lot!

**Comment by jesperancinha** [comment]
I tests this only today. I was following the other branch. Everything works now. Thank you!

**Pull request merged**